File: 938.range-sum-of-bst.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution {
    public $res = 0;

    /**
     * @param TreeNode $root
     * @param Integer $L
     * @param Integer $R
     * @return Integer
     */
    function rangeSumBST($root, $L, $R) {
        $this->dfs($root, $L, $R);
        return $this->res;
    }
    
    function dfs($root, $l, $r) {
        if (empty($root)) {
            return;
        }
        
        if ($root->val < $l) {
            $this->dfs($root->right, $l, $r);
        } else if ($root->val > $r) {
            $this->dfs($root->left, $l, $r);
        } else {
            $this->res += $root->val;
            $this->dfs($root->left, $l, $r);
            $this->dfs($root->right, $l, $r);
        }
    }
}

File: 938.range-sum-of-bst.cpp
// CreateTime: 2019-12-06 16:00:17
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int res = 0;

    int rangeSumBST(TreeNode* root, int L, int R) {
        dfs(root, L, R);
        return res;
    }

    void dfs(TreeNode* root, int l, int r) {
        if (root == NULL) {
            return;
        }

        if (root->val < l) {
            dfs(root->right, l, r);
        } else if (root->val > r) {
            dfs(root->left, l, r);
        } else {
            res += root->val;
            dfs(root->left, l, r);
            dfs(root->right, l, r);
        }
    }
};
